src/deserialize/parse.rs

```diff
@@ -0,0 +1,361 @@
+                .clone()
+                .take_while(|e| e == &'r' || e == &'u' || e == &'e')
+                .collect::<String>()
+                .len();
+            let mut string = String::new();
+            let t = chars.take(c_len).collect::<String>();
+            string.push(c);
+            string.push_str(&t);
+            Ok(Edn::Bool(string.parse::<bool>()?))
+        }
+        'f' => {
+            let c_len = chars
+                .clone()
+                .take_while(|e| e == &'a' || e == &'l' || e == &'s' || e == &'e')
+                .collect::<String>()
+                .len();
+            let mut string = String::new();
+            let f = chars.take(c_len).collect::<String>();
+            string.push(c);
+            string.push_str(&f);
+            Ok(Edn::Bool(string.parse::<bool>()?))
+        }
+        'n' => {
+            let c_len = chars
+                .clone()
+                .take_while(|e| e == &'i' || e == &'l')
+                .collect::<String>()
+                .len();
+            let mut string = String::new();
+            let n = chars.take(c_len).collect::<String>();
+            string.push(c);
+            string.push_str(&n);
+            match &string[..] {
+                "nil" => Ok(Edn::Nil),
+                _ => Err(Error::ParseEdn(format!("{} cound not be parsed", string))),
+            }
+        }
+        _ => Err(Error::ParseEdn(
+            "Nullable boolean cound not be parsed".to_string(),
+        )),
+    }
+}
+
+fn read_vec(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let mut res: Vec<Edn> = vec![];
+    loop {
+        match chars.next() {
+            Some(']') => return Ok(Edn::Vector(Vector::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.push(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
+            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
+        }
+    }
+}
+
+fn read_list(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let mut res: Vec<Edn> = vec![];
+    loop {
+        match chars.next() {
+            Some(')') => return Ok(Edn::List(List::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.push(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
+            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
+        }
+    }
+}
+
+fn read_set(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    use std::collections::BTreeSet;
+    let mut res: BTreeSet<Edn> = BTreeSet::new();
+    loop {
+        match chars.next() {
+            Some('}') => return Ok(Edn::Set(Set::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.insert(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
```

Comment:
Can't this be removed like the other `Some(c)`? :thinking:

Reply:
failures:
    deserialize::parse::test::parse_complex
    deserialize::parse::test::parse_set
    deserialize::test::from_str_complex_map
    deserialize::test::from_str_list_with_set